File: src/Day4.hs
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE ViewPatterns #-}

module Day4 where

import Data.Char (isDigit)
import Data.Maybe (mapMaybe)
import qualified Data.Map as Map
import qualified Data.Set as Set

data Credential = Credential
    { birthYear , issueYear , expirationYear , height , hairColor , eyeColor , passportId :: String
    , countryId :: Maybe String
    }

main :: IO ()
main = do
    credentials <- parseFile <$> readFile "puzzle-inputs/4.txt"
    putStrLn $
        unlines
            [ "Day 4"
            , "Valid passports " <> show (length credentials)
            , "Valid validated passports " <> show (length $ filter isValidCredential credentials)
            ]

parseFile :: String -> [Credential]
parseFile = mapMaybe parseCredential . splitOn mempty . lines

parseCredential :: [String] -> Maybe Credential
parseCredential inp = do
    let fields = Map.fromList $ concatMap (fmap toKeyValue . words) inp
        at = (`Map.lookup` fields)
    birthYear <- at "byr"
    issueYear <- at "iyr"
    expirationYear <- at "eyr"
    height <- at "hgt"
    hairColor <- at "hcl"
    eyeColor <- at "ecl"
    passportId <- at "pid"
    let countryId = at "cid"
    return Credential {..}

toKeyValue :: String -> (String , String)
toKeyValue = (\[key , value] -> (key , value)) . splitOn ':'

isValidCredential :: Credential -> Bool
isValidCredential Credential {..} =
    all
        id
        [ length birthYear == 4 && 1920 <= read @Int birthYear && read @Int birthYear <= 2002
        , length issueYear == 4 && 2010 <= read @Int issueYear && read @Int issueYear <= 2020
        , length expirationYear == 4 && 2020 <= read @Int expirationYear && read @Int expirationYear <= 2030
        , case span isDigit height of
              (read @Int -> n , "in") -> 59 <= n && n <= 76
              (read @Int -> n , "cm") -> 150 <= n && n <= 193
              _ -> False
        , case hairColor of
              '#' : hex -> length hex == 6 && all (`elem` hexNums) hex
              _ -> False
        , eyeColor `Set.member` eyeColors
        , length passportId == 9 && all isDigit passportId
        ]
    where
        hexNums = ['0' .. '9'] ++ ['a' .. 'f']
        eyeColors = Set.fromList ["amb" , "blu" , "brn" , "gry" , "grn" , "hzl" , "oth"]

splitOn :: Ord a => a -> [a] -> [[a]]
splitOn separator l = foldr step [[]] l
    where
        step x (h : acc)
            | x == separator = [] : h : acc
            | otherwise = (x : h) : acc

File: src/day4.rs
use std::collections::HashMap;
use std::fs;
use std::iter::Iterator;

pub fn main() {
    let file = fs::read_to_string("puzzle-inputs/4.txt").expect("Unable to read input file");
    let passports = to_passports(&file);
    println!(
        "Passports valid for criteria 1: {}",
        passports.iter().filter(|p| validate_1(p)).count()
    );
    println!(
        "Passports valid for criteria 2: {}",
        passports
            .iter()
            .filter(|p| REQ_FIELDS.iter().all(|&f| p.contains_key(f)))
            .filter(|p| validate_2(p))
            .count()
    );
}

type Passport = HashMap<String, String>;

fn to_passports(content: &String) -> Vec<Passport> {
    content
        .split("\n\n")
        .map(|line| {
            line.split_ascii_whitespace()
                .map(|w| scan_fmt!(w, "{}:{}", String, String).unwrap())
                .collect()
        })
        .collect()
}

// "cid" is not required
const REQ_FIELDS: [&'static str; 7] = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"];

fn validate_1(p: &Passport) -> bool {
    for &key in REQ_FIELDS.iter() {
        if !p.contains_key(key) {
            return false;
        }
    }
    return true;
}

const EYE_COLORS: [&'static str; 7] = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"];

fn validate_2(p: &Passport) -> bool {
    let checks = [
        p.get("byr")
            .and_then(|v| scan_fmt!(v, "{d}", u32).ok())
            .map(|y| 1920 <= y && y <= 2002),
        p.get("iyr")
            .and_then(|v| scan_fmt!(v, "{d}", u32).ok())
            .map(|y| 2010 <= y && y <= 2020),
        p.get("eyr")
            .and_then(|v| scan_fmt!(v, "{d}", u32).ok())
            .map(|y| 2020 <= y && y <= 2030),
        p.get("hgt").map(|v| {
            if let Ok((n, _)) = scan_fmt!(&v, "{3d}c{[m]}", u32, char) {
                return 150 <= n && n <= 193 && v.len() == 5;
            }
            if let Ok((n, _)) = scan_fmt!(&v, "{2d}i{[n]}", u32, char) {
                return 59 <= n && n <= 76 && v.len() == 4;
            }
            false
        }),
        p.get("hcl")
            .map(|v| scan_fmt!(v, "#{6[0-9a-f]}", String).is_ok()),
        p.get("ecl").map(|v| EYE_COLORS.iter().any(|c| c == v)),
        p.get("pid")
            .map(|v| v.chars().all(|c| c.is_digit(10)) && v.len() == 9),
    ];
    checks.iter().all(|v| v.unwrap_or(false))
}
